dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java

```diff
@@ -30,9 +32,39 @@ public void run(Program program) {
                 obj.getInitialValue(field).accept(collector);
             }
         }
-        // FIXME: We should also traverse the program spec for references to memory objects
+        // Traverse the program spec for references to memory objects
+        addSpecMemoryObjects(program.getSpecification(), collector);
         Sets.difference(memory.getObjects(), collector.memoryObjects).forEach(memory::deleteMemoryObject);
-         */
+    }
+
+    private void addSpecMemoryObjects(AbstractAssert spec, MemoryObjectCollector collector) {
+        if (spec instanceof AssertInline assertInline) {
+            addExprMemoryObject(assertInline.getAssertion().getExpression(), collector);
+        }
+        if (spec instanceof AssertBasic assertBasic) {
+            addExprMemoryObject(assertBasic.getLeft(), collector);
+            addExprMemoryObject(assertBasic.getRight(), collector);
+        }
+        if (spec instanceof AssertNot assertNot) {
+            addSpecMemoryObjects(assertNot.getChild(), collector);
+        }
+        if (spec instanceof AssertCompositeAnd assertCompositeAnd) {
+            addSpecMemoryObjects(assertCompositeAnd.getLeft(), collector);
+            addSpecMemoryObjects(assertCompositeAnd.getRight(), collector);
+        }
+        if (spec instanceof AssertCompositeOr assertCompositeOr) {
+            addSpecMemoryObjects(assertCompositeOr.getLeft(), collector);
+            addSpecMemoryObjects(assertCompositeOr.getRight(), collector);
+        }
+    }
+
+    private void addExprMemoryObject(Expression expr, MemoryObjectCollector collector) {
+        if (expr instanceof Location loc) {
+            loc.getMemoryObject().accept(collector);
+        }
+        if (expr instanceof Register reg) {
+            reg.getFunction().accept(collector);
+        }
```

Comment:
I'm not sure what this was supposed to do, but I think this does not what you intended it to do :).

Reply:
Ahh.. I forgot to delete it